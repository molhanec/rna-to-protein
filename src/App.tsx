import React, { useState } from 'react'
import { Container, Row, Col, FormGroup, Label, ListGroup, ListGroupItem } from 'reactstrap'
import RnaInput from './components/RnaInput'
import ProteinResult from './components/ProteinResult'
import { rnaToProtein, TranslationResult } from "./logic/translate"

function App() {

  const [translationResult, setTranslationResult] = useState<TranslationResult | undefined>(undefined)

  return (
    <Container>
      <Row>
        <Col>
          <h1>RNA to Protein converter</h1>

          <RnaInput
            onTranslate={rna => setTranslationResult(rnaToProtein(rna))}
            invalid={!!(translationResult && translationResult.kind !== 'valid')}
          />

          <ProteinResult protein={
            translationResult?.kind === 'valid' ? translationResult.protein : ''
          } />

          <FormGroup>
            <Label>
              Requirements
            </Label>
            <ListGroup>
              <ListGroupItem>Must start with AUG codon.</ListGroupItem>
              <ListGroupItem>Must end with one of UAA, UAG, or UGA codons.</ListGroupItem>
              <ListGroupItem>UAA, UAG, and UGA codons cannot be in the middle of RNA.</ListGroupItem>
              <ListGroupItem>All codons are triplets.</ListGroupItem>
              <ListGroupItem>Codons consist only of letters A, C, G, and U.</ListGroupItem>
            </ListGroup>
          </FormGroup>

        </Col>
      </Row>
    </Container>
  )
}

export default App
