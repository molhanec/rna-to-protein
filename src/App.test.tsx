import React from 'react'
import { render } from '@testing-library/react'
import App from './App'

describe('<App />', () => {
  it('matches snapshot', () => {
    const { container } = render(<App />)
    expect(container).toMatchSnapshot()
  })
})
