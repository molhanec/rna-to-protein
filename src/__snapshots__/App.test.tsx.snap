// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h1>
          RNA to Protein converter
        </h1>
        <div
          class="form-group"
        >
          <label
            class=""
          >
            RNA codons
          </label>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              type="text"
              value="AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA"
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-primary"
                type="button"
              >
                Translate
              </button>
              <button
                class="btn btn-secondary"
                type="button"
              >
                Reset to example
              </button>
            </div>
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
          >
            Protein
          </label>
          <input
            class="form-control"
            readonly=""
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
          >
            Requirements
          </label>
          <ul
            class="list-group"
          >
            <li
              class="list-group-item"
            >
              Must start with AUG codon.
            </li>
            <li
              class="list-group-item"
            >
              Must end with one of UAA, UAG, or UGA codons.
            </li>
            <li
              class="list-group-item"
            >
              UAA, UAG, and UGA codons cannot be in the middle of RNA.
            </li>
            <li
              class="list-group-item"
            >
              All codons are triplets.
            </li>
            <li
              class="list-group-item"
            >
              Codons consist only of letters A, C, G, and U.
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
